File: leetcode/962. Maximum Width Ramp/main.go
package main

func maxWidthRamp(nums []int) int {
	result := 0
	stack := make([]int, 0, len(nums))

	for i, v := range nums {
		if len(stack) == 0 || nums[stack[len(stack)-1]] > v {
			stack = append(stack, i)
		}
	}

	for j := len(nums) - 1; j >= 0; j-- {
		for len(stack) != 0 {
			i := stack[len(stack)-1]

			if nums[i] <= nums[j] {
				result = max(result, j-i)
				stack = stack[:len(stack)-1]
			} else {
				break
			}
		}
	}

	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}

	return b
}

File: leetcode/962. Maximum Width Ramp/main.rs
impl Solution {
    pub fn max_width_ramp(nums: Vec<i32>) -> i32 {
        let n = nums.len();

        let mut result = 0;
        let mut stack = Vec::new();

        for i in 0..n {
            if stack.is_empty() || nums[*stack.last().unwrap()] > nums[i] {
                stack.push(i);
            }
        }

        for j in (0..n).rev() {
            while let Some(&i) = stack.last() {
                if nums[i] <= nums[j] {
                    result = result.max(j - i);
                    stack.pop();
                } else {
                    break;
                }
            }
        }

        result as i32
    }
}

